torchmetrics/collections.py

```diff
@@ -117,11 +125,83 @@ def update(self, *args: Any, **kwargs: Any) -> None:
+        if self._groups_checked:
+            for _, cg in self._groups.items():
+                m0 = getattr(self, cg[0])
+                m0.update(*args, **m0._filter_kwargs(**kwargs))
+                # copy the state to the remaining metrics in the compute group
+                for i in range(1, len(cg)):
+                    mi = getattr(self, cg[i])
+                    for state in m0._defaults:
+                        setattr(mi, state, getattr(m0, state))
+
+        else:  # the first update always do per metric to form compute groups
+            for _, m in self.items(keep_base=True):
+                m_kwargs = m._filter_kwargs(**kwargs)
+                m.update(*args, **m_kwargs)
+
+            if self._enable_compute_groups:
+                self._merge_compute_groups()
+                self._groups_checked = True
+
+    def _merge_compute_groups(self):
+        """Iterates over the collection of metrics, checking if the state of each metric matches another.
+
+        If so, their compute groups will be merged into one
+        """
+        n_groups = len(self._groups)
+        while True:
+            for cg_idx1, cg_members1 in self._groups.copy().items():
+                for cg_idx2, cg_members2 in self._groups.copy().items():
+                    if cg_idx1 == cg_idx2:
+                        continue
+
+                    metric1 = getattr(self, cg_members1[0])
+                    metric2 = getattr(self, cg_members2[0])
+
+                    if self._equal_metric_states(metric1, metric2):
+                        self._groups[cg_idx1].extend(self._groups.pop(cg_idx2))
+                        break
+
+                # Start over if we merged groups
+                if len(self._groups) != n_groups:
+                    break
+
+            # Stop when we iterate over everything and do not merge any groups
+            if len(self._groups) == n_groups:
+                break
+            else:
+                n_groups = len(self._groups)
+
+    def _equal_metric_states(self, metric1, metric2):
+        """Check if the metric state of two metrics are the same."""
+        if metric1._defaults.keys() != metric2._defaults.keys():
+            return False
+
+        for key in metric1._defaults.keys():
+            state1 = getattr(metric1, key)
+            state2 = getattr(metric2, key)
+
+            if type(state1) != type(state2):
+                return False
+
+            if isinstance(state1, Tensor) and isinstance(state2, Tensor):
+                if state1.shape != state2.shape:
+                    return False
+                if not torch.allclose(state1, state2):
+                    return False
+
+            if isinstance(state1, list) and isinstance(state2, list):
+                if any(s1.shape != s2.shape for s1, s2 in zip(state1, state2)):
+                    return False
+                if any(torch.allclose(s1, s2) for s1, s2 in zip(state1, state2)):
```

Comment:
```suggestion
                if not all(torch.allclose(s1, s2) for s1, s2 in zip(state1, state2)):
```

When moving it to one line as suggested below, this will be outdated anyways, but currently this behaviour is wrong I think :) It either should be `any(not torch.allclose(...))` or `not all(torch.allclose(...))`